src/driver.toit

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2019 Toitware ApS. All rights reserved.
+
+// Driver for Max M8 GPS module.
+
+import monitor
+import serial
+import math
+import reader
+import writer
+import ubx_message
+import gnss_location show GnssLocation
+import location show Location
+
+import .reader
+
+I2C_ADDRESS ::= 0x42
+
+class Driver:
+  static METER_TO_MILLIMETER ::= 1000
+  static METER_TO_CENTIMETER ::= 100
+  static COORDINATE_FACTOR/float ::= 10_000_000.0
+
+  time_to_first_fix_/Duration := Duration
+  waiters_ := []
+
+  location_ /GnssLocation? := null
+  adapter_ /Adapter_
+  runner_ /Task_? := null
+
+  constructor reader/reader.Reader writer --auto_run/bool=true:
+    adapter_ = Adapter_ reader writer
+
+    if auto_run: task --background:: run
+
+  time_to_first_fix: return time_to_first_fix_
+
+  location -> GnssLocation?:
+    return location_
+
+  location --blocking -> GnssLocation:
+    latch := monitor.Latch
+    waiters_.add latch
+    return latch.get
+
+  run:
+    assert: not runner_
+    adapter_.flush
+    start_periodic_nav_packets_
+    runner_ = task::
+      while true:
+        message := adapter_.next_message
+        if message is ubx_message.NavStatus:
+          process_nav_status_ message as ubx_message.NavStatus
+        else if message is ubx_message.NavPvt:
+          process_nav_pvt_ message as ubx_message.NavPvt
+
+  reset:
+    adapter_.reset
+
+  close:
+    if runner_:
+      runner_.cancel
+      runner_ = null
+
+  process_nav_status_ message/ubx_message.NavStatus:
+    if time_to_first_fix_.in_ns != 0: return
+
+    time_to_first_fix_ = Duration --ms=message.time_to_first_fix
+
+  process_nav_pvt_ message/ubx_message.NavPvt:
+    if message.is_gnss_fix:
+      location_ = GnssLocation
+        Location message.lat / COORDINATE_FACTOR message.lon / COORDINATE_FACTOR
```

Comment:
Maybe the GNSS location API is not great here as you need to import both the location package and the GNSS location package to construct a GNSS location.